### PR TITLE
partr: add dynamic scaling and optimize code

### PR DESCRIPTION
**Pull request by vtjnash**
@kpamnany should not see any performance difference here, but solves some of the old TODOs and removes the limitation on `nthreads` being constant

### CODE REVIEW

File: base/partr.jl [hunk 1]
```diff
@@ -61,45 +43,76 @@ function sift_up(heap::taskheap, idx::Int32)
...
+    end
+
+    @lock heaps_lock begin
+        heap_p = UInt32(length(heaps))
+        nt = UInt32(nthreads())
+        if heap_c * nt <= heap_p
+            return heap_p
+        end
+
+        heap_p += heap_c * nt
```
Comment (kpamnany) [inline base/partr.jl:78]: This should not be `+=`. The lowest delay (and also the lowest rank error) for `deletemin` is with `heap_c = 2`; delay and error scale linearly with `heap_c`. We probably do not want `heap_c` to ever exceed 4.

File: base/partr.jl [hunk 2]
```diff
@@ -61,45 +43,76 @@ function sift_up(heap::taskheap, idx::Int32)
...
     end
 end
 
 
+function multiq_size()
+    heap_c = UInt32(2)
+    heap_p = UInt32(length(heaps))
+    nt = UInt32(nthreads())
+
+    if heap_c * nt <= heap_p
```
Comment (kpamnany) [inline base/partr.jl:67]: If `heap_p` is too much larger than `heap_c * nthreads` then `deletemin` delay will be larger than necessary. Suggest leaving a TODO comment here to work out how to resize downwards.

File: base/task.jl [hunk 3]
```diff
@@ -741,18 +747,10 @@ function enq_work(t::Task)
             tid = Threads.threadid()
             ccall(:jl_set_task_tid, Cint, (Any, Cint), t, tid-1)
         end
-        push!(Workqueues[tid], t)
+        push!(workqueue_for(tid), t)
     else
-        if !Partr.multiq_insert(t, t.priority)
-            # if multiq is full, give to a random thread (TODO fix)
```
Comment (kpamnany) [inline base/task.jl:747]: 🎉

File: base/task.jl [hunk 4]
```diff
@@ -708,29 +708,35 @@ function list_deletefirst!(W::IntrusiveLinkedListSynchronized{T}, t::T) where T
...
-        resize!(Workqueues, nt)
-        for i = 2:nt
-            Workqueues[i] = StickyWorkqueue()
+global Workqueues::Vector{StickyWorkqueue} = [StickyWorkqueue()]
+const Workqueues_lock = Threads.SpinLock()
+const Workqueue = Workqueues[1] # default work queue is thread 1 // TODO: deprecate this variable
+
+function workqueue_for(tid::Int)
+    qs = Workqueues
+    if length(qs) >= tid && isassigned(qs, tid)
```
Comment (kpamnany) [inline base/task.jl:717]: Is this `isassigned` call also a performance trap?

**Review by kpamnany — COMMENTED** [review]
There are a couple of different enhancements described in the most recent [`Multiqueues`](https://arxiv.org/pdf/2107.01350.pdf) paper but they do not reduce delay (which is, IMO, the most critical performance criterion for us). However, the buffering enhancement seems to improve throughput without hurting delay which might be worthwhile for us.

**Comment by kpamnany** [comment]
I can rebase https://github.com/JuliaLang/julia/pull/42302 on this (or wait till you merge it first) and ensure that it will be possible to resize the thread pools also. Do we want to also allow dynamically adding/removing the interactive thread pool? That will be trickier.

File: base/partr.jl (hunk 2 above)
Comment (vtjnash) [reply base/partr.jl:67]: I think we might want to alter the scaling to be `min(heap_c * nthreads, current_load)` where `current_load` is the sum of the length of all of the tasks queues (or estimated statistically from a small sample). But I haven't attempted any tuning yet, just focusing on the dynamic scaling for now.

File: base/task.jl (hunk 4 above)
Comment (vtjnash) [reply base/task.jl:717]: this is the correct way to call it

**Comment by vtjnash** [comment]
The buffering enhancement sounds interesting. If I understand the rough idea correctly, we would add a bounded work-stealing queue in front of partr. Work would first go to the local WS queue. When that (small, bounded) queue fills, it dumps the excess work in a batch into partr. You can then select between different algorithms by adjusting the sizes and priorities (pure workstealing: all priorities equal and no bound; pure-partr: no local local WS queue)

Then there is another enhancement I see there, where it tries to re-use the previously selected queue index for some number of consecutive operations, to enhance memory locality of the locking attempts. That also sounds quite interesting.

I think the biggest fix we need here is to change this queue to have, and use, priorities. Currently it is just a queue, and not even a particularly good one (it reshuffles the elements order on every deletemin call)

File: base/partr.jl (hunk 1 above)
Comment (kpamnany) [reply base/partr.jl:78]: Why not just use `=` here?

**Comment by kpamnany** [comment]
It's not work-stealing -- there's no 1:1 (or any sort of) mapping between thread and queue. Their experiments also show that anything greater than size 4 for this buffer will hurt performance.

What the paper calls queue stickiness only benefits throughput, it actually increases delay and error. I didn't see that as useful for us.

The queue does support priorities (that's what the sifting code uses); we aren't setting priorities in tasks (the original parfor implementation did and I think we need it back (with parallel reduction)) because we haven't worked out a coherent scheme to set them. I agree that if we aren't using priorities, this queue is terrible.

File: base/partr.jl (hunk 1 above)
Comment (vtjnash) [reply base/partr.jl:78]: Doesn't scale as well if the thread count is changing dynamically. This initializes to 2x, then averages around 3x if the user is adding threads later.

**Comment by vtjnash** [comment]
The buffers sounded like they are thread-local?

**Comment by kpamnany** [comment]
No, see Sec 4.6. The stickiness counters are thread-local.

**Comment by vtjnash** [comment]
Ah, I was reading the wrong section. The thread-local buffers were in a discussion of related work on CAPQ.

**Pull request merged**